src/recon_buildblock/ML_estimate_component_based_normalisation.cxx

```diff
@@ -18,291 +19,371 @@
  */
 #include "stir/recon_buildblock/ML_estimate_component_based_normalisation.h"
-
 #include "stir/ML_norm.h"
 #include "stir/Scanner.h"
-#include "stir/stream.h"
 #include "stir/display.h"
 #include "stir/info.h"
-#include "stir/warning.h"
 #include "stir/ProjData.h"
-#include <boost/format.hpp>
-#include <fstream>
+#include "stir/stream.h"
 #include <string>
-#include <algorithm>
+#include <utility>
 
 START_NAMESPACE_STIR
 
 void
 ML_estimate_component_based_normalisation(const std::string& out_filename_prefix,
-                                          const ProjData& measured_data,
-                                          const ProjData& model_data,
-                                          int num_eff_iterations,
-                                          int num_iterations,
-                                          bool do_geo,
-                                          bool do_block,
-                                          bool do_symmetry_per_block,
-                                          bool do_KL,
-                                          bool do_display)
+                                          const ProjData& measured_projdata,
+                                          const ProjData& model_projdata,
+                                          const int num_eff_iterations,
+                                          const int num_iterations,
+                                          const bool do_geo,
+                                          const bool do_block,
+                                          const bool do_symmetry_per_block,
+                                          const bool do_KL,
+                                          const bool do_display,
+                                          const bool do_save_to_file)
 {
+  MLEstimateComponentBasedNormalisation estimator(out_filename_prefix,
+                                                  measured_projdata,
+                                                  model_projdata,
+                                                  num_eff_iterations,
+                                                  num_iterations,
+                                                  do_geo,
+                                                  do_block,
+                                                  do_symmetry_per_block,
+                                                  do_KL,
+                                                  do_display,
+                                                  do_save_to_file);
+  estimator.process();
+}
 
-  const int num_transaxial_blocks = measured_data.get_proj_data_info_sptr()->get_scanner_sptr()->get_num_transaxial_blocks();
-  const int num_axial_blocks = measured_data.get_proj_data_info_sptr()->get_scanner_sptr()->get_num_axial_blocks();
+MLEstimateComponentBasedNormalisation::MLEstimateComponentBasedNormalisation(
+    std::string out_filename_prefix_v,
+    const ProjData& measured_projdata_v,
+    const ProjData& model_projdata_v,
```

Comment:
Should there be a check that these ProjData are the same "shape". Are there any tests for that?

Reply:
They should be. Not sure if it's checked anywhere.